crates/net/network/src/session/mod.rs

```diff
@@ -264,6 +265,18 @@ impl SessionManager {
         }
     }
 
+    /// Sends a disconnect message with  TooManyPeers reason.
+    ///
+    /// Called when We currently don't have additional capacity to establish a session from an
+    /// incoming connection.
+    pub(crate) fn disconnect_incoming_connection(&self, stream: TcpStream) {
+        self.spawn(async move {
+            let sink = LengthDelimitedCodec::default().framed(stream);
+            let mut stream = UnauthedP2PStream::new(sink);
```

Comment:
still not sure about the codec, please take a look @Rjected

Reply:
yeah, we should already have an established ECIES stream when sending the disconnect, and `LengthDelimitedCodec` should only be used for testing. So this should probably be done after we build the `UnauthedP2PStream` with the authenticated `ECIESStream`, e.g. here:
https://github.com/paradigmxyz/reth/blob/c4404cc5b8c29df8ca5f5afac3372859d020c16f/crates/net/network/src/session/mod.rs#L784

`send_disconnect` is the right method to use here though. This might need some refactoring of the `authenticate` method, since it doesn't know about incoming capacity

Reply:
so we need to establish a session the way as a pending session?
do we have a way to send the disconnect during the handshake? Because sending the disconnect instead of the hello is valid:

https://github.com/paradigmxyz/reth/blob/d99ba4b4e238dd02ef1d9a0e841dc59cafde2367/crates/net/eth-wire/src/p2pstream.rs#L122-L128

Reply:
> do we have a way to send the disconnect during the handshake?
yes, `UnauthedP2PStream::send_disconnect` would correctly send the disconnect as the first message:
https://github.com/paradigmxyz/reth/blob/c4404cc5b8c29df8ca5f5afac3372859d020c16f/crates/net/eth-wire/src/p2pstream.rs#L75

The remote node will be expecting at least an ECIES handshake, so once we authenticate with ECIES, we can send the disconnect. We could have a method that just establishes the ECIESStream and sends the disconnect when the swarm emits an `ExceedsLimit`. We may not need to add to the pending session map

Reply:
@Rjected Added a commit taking your remarks into account. Can you have a look to be sure I understood correctly? Thanks